subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleResolveState.java

```diff
@@ -302,6 +302,7 @@ public void removeUnattachedDependency(EdgeState edge) {
     }
 
     public ComponentState getVersion(ModuleVersionIdentifier id, ComponentIdentifier componentIdentifier) {
+        assert id.getModule().equals(this.id);
```

Comment:
🧐 `assert` in prod code should be deleted or replaced with a non-disablable `Preconditions` check.

Reply:
Updated. Is the reasoning that we want to be able to check these cases during runtime and not just tests?

Reply:
Preconditions just wraps `if` statements in nice error messages, and can't be disabled.  [Asserts can be](https://softwareengineering.stackexchange.com/questions/137158/is-it-better-to-use-assert-or-illegalargumentexception-for-required-method-param).

Reply:
Seems like the linked post is only really promoting throwing exceptions vs asserts in public methods and not necessarily in internal methods.

Reply:
> It also doesn't throw an exception, it throws an `Error`

This, along with the fact that the assertions are disabled at runtime are compelling arguments. 

> I thought "don't use `assert` except when debugging or in tests" was pretty common advice.

This is what I've traditionally thought as well, though I tend to see a lot of `assert`s in the Gradle codebase which makes me wonder why they're not throwing exceptions. Maybe its more of a sanity check kind of thing? Or thats just the way it is and these should be updated? This class itself has a lot of asserts in it already. 

Either way I'm happy to change this over and will be using proper exception-throwing methods in the future.

Reply:
We use `assert` mostly because they are disabled _by default_ and since the dependency management engine is performance sensitive, this is actually a good thing.
You will find a number of them in this code, their goal is to trip our tests - which run with asserts enabled - in case we end up violating them.
I would make sure that the replacement `Preconditions` is not visible in flame graphs.